File: GetInstance/Extractor.cs
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;

namespace GetInstance
{
	public class Extractor : IExtractor
	{
		IDisplayer displayer;
		public Extractor(IDisplayer displayer)
		{
			this.displayer = displayer;
		}

		public List<WriteFormat> Extract(XDocument xbrl)
		{
			XPathNodeIterator ccRecordsIterator = GetCCIterator(xbrl);
			List<WriteFormat> wf = DeserializeRecords(ccRecordsIterator);
			return wf;
		}
		XPathNodeIterator GetCCIterator(XDocument doc)
		{
			XPathNavigator n = doc.CreateNavigator();
			XmlNamespaceManager m = new XmlNamespaceManager(n.NameTable);
			m.AddNamespace("cc", "http://www.ffiec.gov/xbrl/call/concepts");
			return n.Evaluate("//cc:*", m) as XPathNodeIterator;
		}
		List<WriteFormat> DeserializeRecords(XPathNodeIterator ccElements)
		{
			List<WriteFormat> wf = new();

			while (ccElements.MoveNext())
			{
				var mtContextRef = ccElements.Current.GetAttribute("contextRef", string.Empty);

				if (ccElements.Current.MoveToChild(XPathNodeType.Text))
				{
					bool isConceptDataRecord = bool.TryParse(ccElements.Current.Value, out bool mtBool);
					if (isConceptDataRecord)
					{
						ccElements.Current.MoveToParent();
						displayer.WriteLine(mtContextRef + " -> " + mtBool.ToString());

						wf.Add(new WriteFormat
						{
							MtMdrm = ccElements.Current.LocalName,
							MtData = mtBool ? "true" : "false",
							MtDecimals = "0",
							MtContextRef = mtContextRef
						}) ;
					}
					else
					{
						ccElements.Current.MoveToParent();
						string? mtText = ccElements.Current.Value;
						string? mtMdrm = ccElements.Current.LocalName;
						string? mtUnitRef = ccElements.Current.GetAttribute("unitRef", string.Empty);
						string? mtDecimals = ccElements.Current.GetAttribute("decimals", string.Empty);


						wf.Add(new WriteFormat
						{
							MtDecimals = mtDecimals,
							MtData = mtText,
							MtMdrm = mtMdrm,
							MtUnitRef = mtUnitRef,
							MtContextRef = mtContextRef
						});
					}
				}
				else
				{
					displayer.WriteLine("no text child node?");
				}
			}
			return wf;
		}
	}
}

File: VB/SharedProject1/Extractor.vb
﻿Imports System.Xml
Imports System.Xml.Linq
Imports System.Xml.XPath

Public Class Extractor
	Implements IExtractor
	Private displayer As IDisplayer

		Public Sub New(ByVal displayer As IDisplayer)
			Me.displayer = displayer
		End Sub

		Public Function Extract(ByVal xbrl As XDocument) As List(Of WriteFormat) Implements IExtractor.Extract
			Dim ccRecordsIterator As XPathNodeIterator = GetCCIterator(xbrl)
			Dim wf As List(Of WriteFormat) = DeserializeRecords(ccRecordsIterator)
			Return wf
		End Function

		Private Function GetCCIterator(ByVal doc As XDocument) As XPathNodeIterator
			Dim n As XPathNavigator = doc.CreateNavigator()
			Dim m As XmlNamespaceManager = New XmlNamespaceManager(n.NameTable)
			m.AddNamespace("cc", "http://www.ffiec.gov/xbrl/call/concepts")
			Return TryCast(n.Evaluate("//cc:*", m), XPathNodeIterator)
		End Function

		Private Function DeserializeRecords(ByVal ccElements As XPathNodeIterator) As List(Of WriteFormat)
			Dim wf As List(Of WriteFormat) = New List(Of WriteFormat)

			Dim mtBool As Boolean = Nothing

			While ccElements.MoveNext()
				Dim mtContextRef = ccElements.Current.GetAttribute("contextRef", String.Empty)

				If ccElements.Current.MoveToChild(XPathNodeType.Text) Then
					Dim isConceptDataRecord As Boolean = Boolean.TryParse(ccElements.Current.Value, mtBool)

					If isConceptDataRecord Then
						ccElements.Current.MoveToParent()
						displayer.WriteLine(mtContextRef & " -> " & mtBool.ToString())
						wf.Add(New WriteFormat With {
										.MtMdrm = ccElements.Current.LocalName,
										.MtData = If(mtBool, "true", "false"),
										.MtDecimals = "0",
										.mtContextRef = mtContextRef
						})
					Else
						ccElements.Current.MoveToParent()
						Dim mtText As String = ccElements.Current.Value
						Dim mtMdrm As String = ccElements.Current.LocalName
						Dim mtUnitRef As String = ccElements.Current.GetAttribute("unitRef", String.Empty)
						Dim mtDecimals As String = ccElements.Current.GetAttribute("decimals", String.Empty)
						wf.Add(New WriteFormat With {
										.mtDecimals = mtDecimals,
										.MtData = mtText,
										.mtMdrm = mtMdrm,
										.mtUnitRef = mtUnitRef,
										.mtContextRef = mtContextRef
						})
					End If
				Else
					displayer.WriteLine("no text child node?")
				End If
			End While

			Return wf
		End Function
End Class